src/dtx/dtx_resync.c

```diff
@@ -342,14 +340,28 @@ dtx_resync(daos_handle_t po_hdl, uuid_t po_uuid, uuid_t co_uuid, uint32_t ver,
 		if (!block)
 			goto out;
 
-		shared = true;
-		/* Someone is resyncing the DTXs, let's wait and retry. */
-		ABT_thread_yield();
-	}
+		D_ASSERT(cont->sc_dtx_resync_cbdata == NULL);
+
+		rc = ABT_future_create(1, NULL, &future);
+		if (rc != ABT_SUCCESS) {
+			D_ERROR("ABT_future_create failed for DTX resync on"
+				DF_UUID": rc = %d\n", DP_UUID(co_uuid), rc);
+			D_GOTO(out, rc = dss_abterr2der(rc));
+		}
 
-	/* Some others just did the DTX resync, needs not to repeat. */
-	if (shared)
+		cont->sc_dtx_resync_cbdata = future;
+		/* Someone is resyncing the DTXs, let's wait. */
+		rc = ABT_future_wait(future);
+		D_ASSERTF(rc == ABT_SUCCESS, "ABT_future_wait failed for DTX "
+			  "resync on "DF_UUID" %d\n", DP_UUID(co_uuid), rc);
+
+		cont->sc_dtx_resync_cbdata = NULL;
+		ABT_future_free(&future);
```

Comment:
The ABT_thread_yield() approach is just the original implementation. Xuezhao think it is unnecessary to be busy wait, I agree with him.

Reply:
Not sure what you mean busy wait? future_wait will also wait here? Nothing different as while() { thread_yield()}, besides this is not normal hot path? is it?

Reply:
if dtx_resync is in progress and another 2nd ULT call it again, the original code will cause the 2nd ULT yield and check the sc_dtx_resyncing again and again, forever until the flag cleared by others. This is busy polling and will eat CPU cycle to 100%.

By using future/eventual wait, the 2nd ULT will be suspend by ABT and wake up when the another dtx_resync done and set the future/eventual. This is my understanding.

Reply:
ABT_thread_yield() just yield CPU, and will be rescheduled soon. But ABT_future_wait() will sleep until someone wakeup it when the condition matched.

Reply:
well, if there are no other jobs, just rescheduled it is harmless.  ABT_future_wait() is the same with this regard, IMHO.

Reply:
"This is busy polling and will eat CPU cycle to 100%."

If there are no other jobs, ABT_future_wait will do the same in our cases, since the server will always do busy polling anyway.

Reply:
You mean ABT_future_wait() is also busy polling?

Reply:
I think no, ABT future with a list (p_future->p_head) of ULTs suspended and wait for it.
Although currently overall service XS is busy polling (on cart context), it is better to rely on ABT suspend/wakeup at other places as far as possible.

Reply:
Sorry, I do not see the difference, if there is "nothing" to do, I think ABT_future_wait and while() {ABT_thread_yield()} are same, since ABT will not hang itself anyway, imho

Reply:
If that is true, then it means that the single condition ABT_future_wait() is just equal to "while (!cond) { ABT_thread_yield }" That is some bad implementation.